<a href="https://colab.research.google.com/github/200Richa/Data-Analysis-and-Visualization/blob/main/Lego_Analysis/Lego_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

Today we'll dive deep into a dataset all about LEGO. From the dataset we can ask whole bunch of interesting questions about the history of the LEGO company, their product offering, and which LEGO set ultimately rules them all:

<ul type="square">
<li>What is the most enormous LEGO set ever created and how many parts did it have?</li>

<li>How did the LEGO company start out? In which year were the first LEGO sets released and how many sets did the company sell when it first launched?</li>

<li>Which LEGO theme has the most sets? Is it one of LEGO's own themes like Ninjago or a theme they licensed liked Harry Potter or Marvel Superheroes?</li>

<li>When did the LEGO company really expand its product offering? Can we spot a change in the company strategy based on how many themes and sets did it released year-on-year?</li>

<li>Did LEGO sets grow in size and complexity over time? Do older LEGO 
sets tend to have more or fewer parts than newer sets?</li>
</ul>

**Data Source**

[Rebrickable](https://rebrickable.com/downloads/) has compiled data on all the LEGO pieces in existence. I recommend you use download the .csv files provided in this lesson. 

<img src="https://i.imgur.com/49FNOHj.jpg">

# Import Statements

In [41]:
import pandas as pd
import matplotlib.pyplot as plt

# Data Exploration

**Question**: How many different colours does the LEGO company produce? Read the colors.csv file in the data folder and find the total number of unique colours. Try using the [.nunique() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.nunique.html?highlight=nunique#pandas.DataFrame.nunique) to accomplish this.

In [42]:
colors = pd.read_csv('data/colors.csv')

In [43]:
colors

id                          name     rgb is_trans
0      -1                       Unknown  0033B2        f
1       0                         Black  05131D        f
2       1                          Blue  0055BF        f
3       2                         Green  237841        f
4       3                Dark Turquoise  008F9B        f
..    ...                           ...     ...      ...
130  1004  Trans Flame Yellowish Orange  FCB76D        t
131  1005             Trans Fire Yellow  FBE890        t
132  1006        Trans Light Royal Blue  B4D4F7        t
133  1007                 Reddish Lilac  8E5597        f
134  9999                    [No Color]  05131D        f

[135 rows x 4 columns]

In [44]:
colors.head()

id            name     rgb is_trans
0  -1         Unknown  0033B2        f
1   0           Black  05131D        f
2   1            Blue  0055BF        f
3   2           Green  237841        f
4   3  Dark Turquoise  008F9B        f

In [45]:
colors.tail()

id                          name     rgb is_trans
130  1004  Trans Flame Yellowish Orange  FCB76D        t
131  1005             Trans Fire Yellow  FBE890        t
132  1006        Trans Light Royal Blue  B4D4F7        t
133  1007                 Reddish Lilac  8E5597        f
134  9999                    [No Color]  05131D        f

In [46]:
colors.shape

(135, 4)

In [47]:
colors.nunique()

id          135
name        135
rgb         124
is_trans      2
dtype: int64

In [48]:
colors.nunique(axis=1)

0      4
1      4
2      4
3      4
4      4
      ..
130    4
131    4
132    4
133    4
134    4
Length: 135, dtype: int64

In [49]:
colors.columns

Index(['id', 'name', 'rgb', 'is_trans'], dtype='object')

In [50]:
colors.nunique()['name']

135

In [51]:
colors['name'].nunique()

135

In [52]:
# f"There are {colors.nunique().rgb} unique colors lego offers"
f"There are {colors.nunique()['name']} unique colors lego offers"

'There are 135 unique colors lego offers'

**Question**: Find the number of transparent colours where <code>is_trans == 't'</code> versus the number of opaque colours where <code>is_trans == 'f'</code>. See if you can accomplish this in two different ways.

In [53]:
colors.groupby('is_trans').count()

id  name  rgb
is_trans                
f         107   107  107
t          28    28   28

In [54]:
colors.is_trans.value_counts()

f    107
t     28
Name: is_trans, dtype: int64

In [55]:
colors.is_trans.value_counts()['f']

107

In [56]:
transparent = colors.is_trans[colors.is_trans == 't'].count()
transparent

28

In [57]:
opaque = colors.is_trans[colors.is_trans == 'f'].count()
opaque

107

### Understanding LEGO Themes vs. LEGO Sets
Walk into a LEGO store and you will see their products organised by theme. Their themes include Star Wars, Batman, Harry Potter and many more.

<img src="https://i.imgur.com/aKcwkSx.png" >

A lego **set** is a particular box of LEGO or product. Therefore, a single theme typically has many different sets.

<img src="https://i.imgur.com/whB1olq.png" > 

The <code>sets.csv</code> data contains a list of sets over the years and the number of parts that each of these sets contained.

**Question**: Read the sets.csv data and take a look at the first and last couple of rows. 

In [59]:
sets = pd.read_csv('data/sets.csv')

In [60]:
sets.head()

set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1978        84         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12

In [61]:
sets.tail()

set_num                                 name  year  theme_id  \
15705      wwgp1-1  Wild West Limited Edition Gift Pack  1996       476   
15706   XMASTREE-1                       Christmas Tree  2019       410   
15707      XWING-1                  Mini X-Wing Fighter  2019       158   
15708      XWING-2                    X-Wing Trench Run  2019       158   
15709  YODACHRON-1      Yoda Chronicles Promotional Set  2013       158   

       num_parts  
15705          0  
15706         26  
15707         60  
15708         52  
15709        413

In [62]:
sets.shape

(15710, 5)

**Question**: In which year were the first LEGO sets released and what were these sets called?

In [69]:
first_year = sets['year'].min()
f"The first lego set was released in {sets['year'].min()}"

'The first lego set was released in 1949'

In [68]:
f"These sets were called {sets.loc[sets['year'].idxmin()]['name']}"

'These sets were called Extra-Large Gift Set (ABB)'

**Question**: How many different sets did LEGO sell in their first year? How many types of LEGO products were on offer in the year the company started?

In [70]:
sets.head()

set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1978        84         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12

In [73]:
diff_sets =sets.year.value_counts()[first_year]
f"LEGO sell {diff_sets} different sets in their first year"

'LEGO sell 5 different sets in their first year'

**Question**: Find the top 5 LEGO sets with the most number of parts. 

In [74]:
sets.head()

set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1978        84         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12

In [77]:
sets_parts = sets.sort_values('num_parts', ascending=False)
sets_parts

set_num                                              name  year  \
15004   BIGBOX-1                     The Ultimate Battle for Chima  2015   
11183    75192-1                             UCS Millennium Falcon  2017   
10551    71043-1                                   Hogwarts Castle  2018   
295      10256-1                                         Taj Mahal  2017   
221      10189-1                                         Taj Mahal  2008   
...          ...                                               ...   ...   
1782     20216-1  MBA Robot & Micro Designer (Kits 2 - 3 Redesign)  2013   
1780     20214-1      MBA Adventure Designer (Kits 7 - 9 Redesign)  2013   
6822   5005539-1                              Brick Pouch (Yellow)  2018   
9026     66319-1                     Power Miners 3 in 1 Superpack  2009   
12946   853471-1                                   C-3PO Key Chain  2015   

       theme_id  num_parts  
15004       571       9987  
11183       171       7541  
10551       246       6020  
295         673       5923  
221         673       5922  
...         ...        ...  
1782        432          0  
1780        432          0  
6822        501          0  
9026        439          0  
12946       503          0  

[15710 rows x 5 columns]

In [78]:
sets_parts.head()

set_num                           name  year  theme_id  num_parts
15004  BIGBOX-1  The Ultimate Battle for Chima  2015       571       9987
11183   75192-1          UCS Millennium Falcon  2017       171       7541
10551   71043-1                Hogwarts Castle  2018       246       6020
295     10256-1                      Taj Mahal  2017       673       5923
221     10189-1                      Taj Mahal  2008       673       5922

The largest LEGO set ever produced has around 10,000 pieces! Apparently, only two of these boxes were ever produced, so if you wanted to get your hands on a ridiculously large LEGO set, you'll have to settle for the 7,541 piece Millennium Falcon. 


<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-10_10-03-07-e8dce27efa5da9efb1c6fd8219c5ae53.jpeg" >

**Challenge**: Use <code>.groupby()</code> and <code>.count()</code> to show the number of LEGO sets released year-on-year. How do the number of sets released in 1955 compare to the number of sets released in 2019? 

In [79]:
sets.groupby('year').count()

set_num  name  theme_id  num_parts
year                                    
1949        5     5         5          5
1950        6     6         6          6
1953        4     4         4          4
1954       14    14        14         14
1955       28    28        28         28
...       ...   ...       ...        ...
2017      786   786       786        786
2018      816   816       816        816
2019      840   840       840        840
2020      674   674       674        674
2021        3     3         3          3

[71 rows x 4 columns]

**Challenge**: Show the number of LEGO releases on a line chart using Matplotlib. <br>
<br>
Note that the .csv file is from late 2020, so to plot the full calendar years, you will have to exclude some data from your chart. Can you use the slicing techniques covered in Day 21 to avoid plotting the last two years? The same syntax will work on Pandas DataFrames. 

### Aggregate Data with the Python .agg() Function

Let's work out the number of different themes shipped by year. This means we have to count the number of unique theme_ids per calendar year.

**Challenge**: Plot the number of themes released by year on a line chart. Only include the full calendar years (i.e., exclude 2020 and 2021). 

### Line Charts with Two Seperate Axes

**Challenge**: Use the <code>.groupby()</code> and <code>.agg()</code> function together to figure out the average number of parts per set. How many parts did the average LEGO set released in 1954 compared to say, 2017?

### Scatter Plots in Matplotlib

**Challenge**: Has the size and complexity of LEGO sets increased over time based on the number of parts? Plot the average number of parts over time using a Matplotlib scatter plot. See if you can use the [scatter plot documentation](https://matplotlib.org/3.1.0/api/_as_gen/matplotlib.pyplot.scatter.html) before I show you the solution. Do you spot a trend in the chart? 

### Number of Sets per LEGO Theme

LEGO has licensed many hit franchises from Harry Potter to Marvel Super Heros to many others. But which theme has the largest number of individual sets? 

**Challenge** Use what you know about HTML markup and tags to display the database schema: https://i.imgur.com/Sg4lcjx.png

### Database Schemas, Foreign Keys and Merging DataFrames

The themes.csv file has the actual theme names. The sets .csv has <code>theme_ids</code> which link to the <code>id</code> column in the themes.csv. 

**Challenge**: Explore the themes.csv. How is it structured? Search for the name 'Star Wars'. How many <code>id</code>s correspond to this name in the themes.csv? Now use these <code>id</code>s and find the corresponding the sets in the sets.csv (Hint: you'll need to look for matches in the <code>theme_id</code> column)

### Merging (i.e., Combining) DataFrames based on a Key
